hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/fair/FSAppAttempt.java

```diff
@@ -831,7 +831,11 @@ private Resource assignContainer(
     }
 
     // The desired container won't fit here, so reserve
-    if (isReservable(capability) &&
+    // Important: Do not enable this kind of reservation in case of preemption,
+    // since the preempted application will re-reserve the preempted resources
+    // and prevent the starving application to get them
+    if (!scheduler.getConf().getPreemptionEnabled() &&
```

Comment:
I improved this by giving higher priority to preempted containers than reservation. Child queues may pass containers to each other this way resolving fairness starvation. Please review the updated change.